RFC: Segregate binaries created by deps out of usr/bin

tkelman opened the pull request:
since people are apparently adding this location to their path, we should remove
build-time-only clutter from it

I may need to adjust some things on appveyor ~~(and for windows source builds)~~ for this to go through.

todo:
- [x] get working for cmake-built llvm

tkelman committed:
Segregate binaries created by deps out of usr/bin

since people are apparently adding this location to their path,
we should remove build-time-only clutter from it

Also set shlibdir in MAKE_COMMON so openlibm and openspecfun don't
install dlls into usr/tools (build_depsbindir) on windows

Add a patch to customize bindir in cmake-build llvm
using CMAKE_INSTALL_BINDIR variable
set CMAKE_INSTALL_BINDIR in cross compile NATIVE.cmake toolchain file

also copy gcc dlls to usr/tools on windows
split std_dll into 2 rules for build_bindir and build_depsbindir

ViralBShah commented:
+1

tkelman commented:
cc @Keno this may have consequences for the cxx build configuration? This moves things out of `usr/bin` that wouldn't be present in `bin` in a binary. My understanding is Cxx is not currently capable of running in a binary build of Julia, right?

Keno commented:
Correct, it needs to access private clang APIs, so it needs to build it's own shared library that has clang and it's own code to access those.

tkelman commented:
Do you need `llvm-config` or any of the other llvm executables in order to do so? If we were to add a shared libclang to the binaries, either by default or in a custom cxx configuration (which I think we've been trying to do off-and-on for a while?), would that be enough?

Keno commented:
It does look at llvm-config for system libs `https://github.com/Keno/Cxx.jl/blob/master/deps/BuildBootstrap.Makefile#L40`. Shared libclang is unfortunately not enough because it needs private APIs.

tkelman commented:
Hm okay. I don't think we will generally want to ship an `llvm-config` in a binary install of Julia, or if we do it should be separated in a different location than the julia binary to avoid the path issue. Let me know if anything here needs to change for Cxx in the existing source build setup where it does work.

tkelman commented:
Since you're including Make.inc, I think it might just work. I don't have a Cxx compatible build handy to test on though.

tkelman commented:
If I don't hear otherwise, I'll merge this later this week.

vtjnash commented:
the main concern i have is that this requires rebuilding all of deps, and I don't see a good way to do that (short of using a new name for the in-tree `deps/build` folder).

tkelman commented:
should only require re-running configure and reinstalling?

ViralBShah commented:
Given that we no longer do `make check`, it is not so bad to rebuild deps, I feel.

tkelman commented:
It doesn't actually need to recompile any of them, as long as it re-does configure for patchelf and `make install` to copy files around for the rest. Calling `make cleanall` will force the latter.

tkelman commented:
The patch here for making this work when you build LLVM using cmake is maybe not the cleanest. I'll bring this up with upstream, maybe ask on IRC for some initial feedback if people think this kind of customization of `bindir` would be desired before opening a review on phabricator.

tkelman committed:
Upgrade patchelf to 0.9 and mpfr to 3.1.4

tkelman commented:
Rebased and ready, Keno or Jameson any objections to doing this? llvm and other build-only tools would now be in `usr/tools` instead of `usr/bin`. Updating existing build trees after merging this would need to apply the new patch to llvm and recompile in place, but updating patchelf at the same time helps avoid any other migration issues. None of the other deps should need rebuilding, other than an mpfr version bump which is small.

LLVM patch will be upstreamed if they'll take it.

vtjnash commented on Makefile:
```diff
@@ -298,8 +298,9 @@ endif
 julia-deps: | $$(build_bindir)/lib$(1).dll
-$$(build_bindir)/lib$(1).dll: | $$(build_bindir)
-	cp $$(call pathsearch,lib$(1).dll,$$(STD_LIB_PATH)) $$(build_bindir) ;
+$$(build_bindir)/lib$(1).dll: | $$(build_bindir) $$(build_depsbindir)
```
a target should only do exactly one thing, not make two files. make this two targets and add both to the julia-deps target on line 300

vtjnash commented on deps/Makefile:
```diff
@@ -79,7 +79,7 @@ endif
 # If the top-level Makefile is called with environment variables,
 # they will override the values passed above to ./configure
-MAKE_COMMON := DESTDIR="" prefix=$(build_prefix) bindir=$(build_bindir) libdir=$(build_libdir) libexecdir=$(build_libexecdir) datarootdir=$(build_datarootdir) includedir=$(build_includedir) sysconfdir=$(build_sysconfdir) O=
+MAKE_COMMON := DESTDIR="" prefix=$(build_prefix) libdir=$(build_libdir) libexecdir=$(build_libexecdir) datarootdir=$(build_datarootdir) includedir=$(build_includedir) sysconfdir=$(build_sysconfdir) O=
```
leave bindir here, it's preventing `make bindir=` from propagating to the dependencies unintentionally

tkelman replied:
that's a little broken since some of the makefile-only deps are using bindir for dlls. will check again if it's only the ones easily under our control. see commit message

> Do not override bindir in MAKE_COMMON since only autotools-using deps
install any executables, and some of the makefile-only deps (openlibm,
openspecfun) install dlls into bindir on windows so we don't want to
set that to usr/tools

vtjnash commented on deps/llvm-3.7.1_tools.patch:
```diff
@@ -0,0 +1,58 @@
+     set(EXCLUDE_FROM_ALL ON)
+   endif()
+   add_llvm_executable(${name} ${ARGN})
++  if( NOT DEFINED LLVM_TOOLS_INSTALL_DIR )
```
can llvm use the standard names for this instead of inventing our own? (https://cmake.org/cmake/help/v3.0/module/GNUInstallDirs.html)

tkelman replied:
d'oh, how did I miss those. That should be much simpler.

tkelman commented:
Review addressed. (https://github.com/tkelman/julia/commit/d578c9ecfa4503c2ff7f0edc75eb3b9af9c8d2c5 for non-squashed version of diff)

tkelman committed:
Patch for 3.7.1 was not working with symlinks

instead, backport a series of llvm commits so the 3.8 version of the bindir
patch also works

dont try to delete a file in the patch, it should be harmless to leave it

tkelman commented:
This is ready as far as I'm concerned. Note that I changed the cmake-built llvm symlink handling approach to just backport the series of cmake fixes from 3.8 that made the whole thing work better.

Since this and #16723 both add llvm patches, they would be convenient to merge close to one another to minimize how many times people have to recompile llvm with new patches applied. There may be a slight conflict in the list of llvm patches, but both PR's are strictly adding and don't need to add any cross-dependencies since one is patching cmake and the other is patching makefiles.
